File: wearapp/app/src/main/java/com/benoitletondor/pixelminimalwatchface/ComplicationConfigRecyclerViewAdapter.java
package com.benoitletondor.pixelminimalwatchface;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.support.wearable.complications.ComplicationHelperActivity;
import android.support.wearable.complications.ComplicationProviderInfo;
import android.support.wearable.complications.ProviderInfoRetriever;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.concurrent.Executors;

public class ComplicationConfigRecyclerViewAdapter
        extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    public enum ComplicationLocation {
        LEFT,
        RIGHT
    }

    public static final int TYPE_PREVIEW_AND_COMPLICATIONS_CONFIG = 0;
    public static final int TYPE_COLOR_CONFIG = 2;

    // ComponentName associated with watch face service (service that renders watch face). Used
    // to retrieve complication information.
    private ComponentName mWatchFaceComponentName;

    private ArrayList<ComplicationConfigData.ConfigItemType> mSettingsDataSet;

    private Context mContext;

    // Selected complication id by user.
    private int mSelectedComplicationId;

    private int mLeftComplicationId;
    private int mRightComplicationId;

    // Required to retrieve complication data from watch face for preview.
    private ProviderInfoRetriever mProviderInfoRetriever;

    // Maintains reference view holder to dynamically update watch face preview. Used instead of
    // notifyItemChanged(int position) to avoid flicker and re-inflating the view.
    private PreviewAndComplicationsViewHolder mPreviewAndComplicationsViewHolder;

    public ComplicationConfigRecyclerViewAdapter(
            Context context,
            Class watchFaceServiceClass,
            ArrayList<ComplicationConfigData.ConfigItemType> settingsDataSet) {

        mContext = context;
        mWatchFaceComponentName = new ComponentName(mContext, watchFaceServiceClass);
        mSettingsDataSet = settingsDataSet;

        // Default value is invalid (only changed when user taps to change complication).
        mSelectedComplicationId = -1;

        mLeftComplicationId =
                PixelMinimalWatchFace.Companion.getComplicationId(ComplicationLocation.LEFT);
        mRightComplicationId =
                PixelMinimalWatchFace.Companion.getComplicationId(ComplicationLocation.RIGHT);

        // Initialization of code to retrieve active complication data for the watch face.
        mProviderInfoRetriever =
                new ProviderInfoRetriever(mContext, Executors.newCachedThreadPool());
        mProviderInfoRetriever.init();
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        RecyclerView.ViewHolder viewHolder = null;

        switch (viewType) {
            case TYPE_PREVIEW_AND_COMPLICATIONS_CONFIG:
                // Need direct reference to watch face preview view holder to update watch face
                // preview based on selections from the user.
                mPreviewAndComplicationsViewHolder =
                        new PreviewAndComplicationsViewHolder(
                                LayoutInflater.from(parent.getContext())
                                        .inflate(
                                                R.layout.config_list_preview_and_complications_item,
                                                parent,
                                                false));
                viewHolder = mPreviewAndComplicationsViewHolder;
                break;

            case TYPE_COLOR_CONFIG:
                viewHolder =
                        new ColorPickerViewHolder(
                                LayoutInflater.from(parent.getContext())
                                        .inflate(R.layout.config_list_color_item, parent, false));
                break;
        }

        return viewHolder;
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder viewHolder, int position) {
        // Pulls all data required for creating the UX for the specific setting option.
        ComplicationConfigData.ConfigItemType configItemType = mSettingsDataSet.get(position);

        switch (viewHolder.getItemViewType()) {
            case TYPE_PREVIEW_AND_COMPLICATIONS_CONFIG:
                PreviewAndComplicationsViewHolder previewAndComplicationsViewHolder =
                        (PreviewAndComplicationsViewHolder) viewHolder;

                ComplicationConfigData.PreviewAndComplicationsConfigItem previewAndComplicationsConfigItem =
                        (ComplicationConfigData.PreviewAndComplicationsConfigItem) configItemType;

                int defaultComplicationResourceId =
                        previewAndComplicationsConfigItem.getDefaultComplicationResourceId();
                previewAndComplicationsViewHolder.setDefaultComplicationDrawable(
                        defaultComplicationResourceId);

                previewAndComplicationsViewHolder.initializesColorsAndComplications();
                break;

            case TYPE_COLOR_CONFIG:
                ColorPickerViewHolder colorPickerViewHolder = (ColorPickerViewHolder) viewHolder;
                ComplicationConfigData.ColorConfigItem colorConfigItem = (ComplicationConfigData.ColorConfigItem) configItemType;

                int iconResourceId = colorConfigItem.getIconResourceId();
                String name = colorConfigItem.getName();
                Class<ColorSelectionActivity> activity =
                        colorConfigItem.getActivityToChoosePreference();

                colorPickerViewHolder.setIcon(iconResourceId);
                colorPickerViewHolder.setName(name);
                colorPickerViewHolder.setLaunchActivityToSelectColor(activity);
                break;
        }
    }

    @Override
    public int getItemViewType(int position) {
        ComplicationConfigData.ConfigItemType configItemType = mSettingsDataSet.get(position);
        return configItemType.getConfigType();
    }

    @Override
    public int getItemCount() {
        return mSettingsDataSet.size();
    }

    /** Updates the selected complication id saved earlier with the new information. */
    public void updateSelectedComplication(ComplicationProviderInfo complicationProviderInfo) {
        // Checks if view is inflated and complication id is valid.
        if (mPreviewAndComplicationsViewHolder != null && mSelectedComplicationId >= 0) {
            mPreviewAndComplicationsViewHolder.updateComplicationViews(
                    mSelectedComplicationId, complicationProviderInfo,
                    Storage.INSTANCE.getComplicationColors()
            );
        }
    }

    @Override
    public void onDetachedFromRecyclerView(@NonNull RecyclerView recyclerView) {
        super.onDetachedFromRecyclerView(recyclerView);
        // Required to release retriever for active complication data on detach.
        mProviderInfoRetriever.release();
    }

    public void updatePreviewColors() {
        mPreviewAndComplicationsViewHolder.updateComplicationsAccentColor(Storage.INSTANCE.getComplicationColors());
    }

    /**
     * Displays watch face preview along with complication locations. Allows user to tap on the
     * complication they want to change and preview updates dynamically.
     */
    public class PreviewAndComplicationsViewHolder extends RecyclerView.ViewHolder
            implements OnClickListener {

        private ImageView mLeftComplicationBackground;
        private ImageView mRightComplicationBackground;

        private ImageButton mLeftComplication;
        private ImageButton mRightComplication;

        private Drawable mDefaultComplicationDrawable;

        public PreviewAndComplicationsViewHolder(final View view) {
            super(view);

            // Sets up left complication preview.
            mLeftComplicationBackground = view.findViewById(R.id.left_complication_background);
            mLeftComplication = view.findViewById(R.id.left_complication);
            mLeftComplication.setOnClickListener(this);

            // Sets up right complication preview.
            mRightComplicationBackground = view.findViewById(R.id.right_complication_background);
            mRightComplication = view.findViewById(R.id.right_complication);
            mRightComplication.setOnClickListener(this);
        }

        public void setDefaultComplicationDrawable(int resourceId) {
            Context context = mLeftComplicationBackground.getContext();
            mDefaultComplicationDrawable = context.getDrawable(resourceId);

            mLeftComplication.setImageDrawable(mDefaultComplicationDrawable);
            mLeftComplicationBackground.setVisibility(View.INVISIBLE);

            mRightComplication.setImageDrawable(mDefaultComplicationDrawable);
            mRightComplicationBackground.setVisibility(View.INVISIBLE);
        }

        @Override
        public void onClick(View view) {
            if (view.equals(mLeftComplication)) {
                Activity currentActivity = (Activity) view.getContext();
                launchComplicationHelperActivity(currentActivity, ComplicationLocation.LEFT);
            } else if (view.equals(mRightComplication)) {
                Activity currentActivity = (Activity) view.getContext();
                launchComplicationHelperActivity(currentActivity, ComplicationLocation.RIGHT);
            }
        }

        // Verifies the watch face supports the complication location, then launches the helper
        // class, so user can choose their complication data provider.
        private void launchComplicationHelperActivity(
                Activity currentActivity, ComplicationLocation complicationLocation) {

            mSelectedComplicationId = PixelMinimalWatchFace.Companion.getComplicationId(complicationLocation);

            if (mSelectedComplicationId >= 0) {

                int[] supportedTypes =
                        PixelMinimalWatchFace.Companion.getSupportedComplicationTypes(
                                complicationLocation);

                ComponentName watchFace =
                        new ComponentName(
                                currentActivity, PixelMinimalWatchFace.class);

                currentActivity.startActivityForResult(
                        ComplicationHelperActivity.createProviderChooserHelperIntent(
                                currentActivity,
                                watchFace,
                                mSelectedComplicationId,
                                supportedTypes),
                        ComplicationConfigActivity.COMPLICATION_CONFIG_REQUEST_CODE);

            }
        }

        public void updateComplicationViews(
                int watchFaceComplicationId, ComplicationProviderInfo complicationProviderInfo, @NonNull ComplicationColors complicationColors) {

            if (watchFaceComplicationId == mLeftComplicationId) {
                updateComplicationView(complicationProviderInfo, mLeftComplication,
                        mLeftComplicationBackground, complicationColors);

            } else if (watchFaceComplicationId == mRightComplicationId) {
                updateComplicationView(complicationProviderInfo, mRightComplication,
                        mRightComplicationBackground, complicationColors);
            }
        }

        private void updateComplicationView(ComplicationProviderInfo complicationProviderInfo,
                                            ImageButton button,
                                            ImageView background,
                                            @NonNull ComplicationColors complicationColors) {
            if (complicationProviderInfo != null) {
                button.setImageIcon(complicationProviderInfo.providerIcon);
                button.setContentDescription(
                        mContext.getString(R.string.edit_complication,
                                complicationProviderInfo.appName + " " +
                                        complicationProviderInfo.providerName));
                background.setVisibility(View.VISIBLE);
            } else {
                button.setImageDrawable(mDefaultComplicationDrawable);
                button.setContentDescription(mContext.getString(R.string.add_complication));
                background.setVisibility(View.INVISIBLE);
            }

            updateComplicationsAccentColor(complicationColors);
        }

        public void initializesColorsAndComplications() {

            final int[] complicationIds = PixelMinimalWatchFace.Companion.getComplicationIds();

            mProviderInfoRetriever.retrieveProviderInfo(
                    new ProviderInfoRetriever.OnProviderInfoReceivedCallback() {
                        @Override
                        public void onProviderInfoReceived(
                                int watchFaceComplicationId,
                                @Nullable ComplicationProviderInfo complicationProviderInfo) {
                            updateComplicationViews(
                                    watchFaceComplicationId, complicationProviderInfo, Storage.INSTANCE.getComplicationColors());
                        }
                    },
                    mWatchFaceComponentName,
                    complicationIds);
        }

        public void updateComplicationsAccentColor(@NonNull ComplicationColors colors) {
            mRightComplication.setColorFilter(colors.getRightColor());
            mLeftComplication.setColorFilter(colors.getLeftColor());
        }
    }

    /**
     * Displays color options for the an item on the watch face. These could include marker color,
     * background color, etc.
     */
    public class ColorPickerViewHolder extends RecyclerView.ViewHolder implements OnClickListener {

        private Button mAppearanceButton;

        private Class<ColorSelectionActivity> mLaunchActivityToSelectColor;

        public ColorPickerViewHolder(View view) {
            super(view);

            mAppearanceButton = view.findViewById(R.id.color_picker_button);
            view.setOnClickListener(this);
        }

        public void setName(String name) {
            mAppearanceButton.setText(name);
        }

        public void setIcon(int resourceId) {
            Context context = mAppearanceButton.getContext();
            mAppearanceButton.setCompoundDrawablesWithIntrinsicBounds(
                    context.getDrawable(resourceId), null, null, null);
        }

        public void setLaunchActivityToSelectColor(Class<ColorSelectionActivity> activity) {
            mLaunchActivityToSelectColor = activity;
        }

        @Override
        public void onClick(View view) {
            int position = getAdapterPosition();

            if (mLaunchActivityToSelectColor != null) {
                Intent launchIntent = new Intent(view.getContext(), mLaunchActivityToSelectColor);

                Activity activity = (Activity) view.getContext();
                activity.startActivityForResult(
                        launchIntent,
                        ComplicationConfigActivity.UPDATE_COLORS_CONFIG_REQUEST_CODE);
            }
        }
    }
}

File: wearapp/app/src/main/java/com/benoitletondor/pixelminimalwatchface/ComplicationConfigRecyclerViewAdapter.kt
package com.benoitletondor.pixelminimalwatchface

import android.app.Activity
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.graphics.drawable.Drawable
import android.support.wearable.complications.ComplicationHelperActivity
import android.support.wearable.complications.ComplicationProviderInfo
import android.support.wearable.complications.ProviderInfoRetriever
import android.support.wearable.complications.ProviderInfoRetriever.OnProviderInfoReceivedCallback
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import androidx.recyclerview.widget.RecyclerView
import com.benoitletondor.pixelminimalwatchface.PixelMinimalWatchFace.Companion.getComplicationId
import com.benoitletondor.pixelminimalwatchface.PixelMinimalWatchFace.Companion.getComplicationIds
import com.benoitletondor.pixelminimalwatchface.PixelMinimalWatchFace.Companion.getSupportedComplicationTypes
import java.util.concurrent.Executors

private const val TYPE_PREVIEW_AND_COMPLICATIONS_CONFIG = 0
private const val TYPE_COLOR_CONFIG = 1

class ComplicationConfigRecyclerViewAdapter(
    private val mContext: Context,
    private val storage: Storage
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    enum class ComplicationLocation {
        LEFT, RIGHT
    }

    private val mWatchFaceComponentName: ComponentName = ComponentName(mContext, PixelMinimalWatchFace::class.java)

    private var mSelectedComplicationLocation: ComplicationLocation? = null

    private val mLeftComplicationId: Int = getComplicationId(ComplicationLocation.LEFT)
    private val mRightComplicationId: Int =  getComplicationId(ComplicationLocation.RIGHT)

    private val mProviderInfoRetriever: ProviderInfoRetriever = ProviderInfoRetriever(mContext, Executors.newCachedThreadPool())

    private lateinit var mPreviewAndComplicationsViewHolder: PreviewAndComplicationsViewHolder

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        when (viewType) {
            TYPE_PREVIEW_AND_COMPLICATIONS_CONFIG -> {
                mPreviewAndComplicationsViewHolder = PreviewAndComplicationsViewHolder(
                    LayoutInflater.from(parent.context)
                        .inflate(
                            R.layout.config_list_preview_and_complications_item,
                            parent,
                            false
                        ),
                    object : ComplicationClickListener {
                        override fun onComplicationClicked(location: ComplicationLocation) {
                            mSelectedComplicationLocation = location
                            val selectedComplicationId = getComplicationId(location)
                            if (selectedComplicationId >= 0) {
                                val supportedTypes = getSupportedComplicationTypes(location)

                                val watchFace = ComponentName(mContext, PixelMinimalWatchFace::class.java)
                                (mContext as Activity).startActivityForResult(
                                    ComplicationHelperActivity.createProviderChooserHelperIntent(
                                        mContext,
                                        watchFace,
                                        selectedComplicationId,
                                        *supportedTypes
                                    ),
                                    ComplicationConfigActivity.COMPLICATION_CONFIG_REQUEST_CODE
                                )
                            }
                        }
                    }
                )

                return mPreviewAndComplicationsViewHolder!!
            }
            TYPE_COLOR_CONFIG -> return ColorPickerViewHolder(
                LayoutInflater.from(parent.context).inflate(
                    R.layout.config_list_color_item,
                    parent,
                    false
                )
            )
        }
        throw IllegalStateException("Unknown option type: $viewType")
    }

    override fun onBindViewHolder(viewHolder: RecyclerView.ViewHolder, position: Int) {
        when (viewHolder.itemViewType) {
            TYPE_PREVIEW_AND_COMPLICATIONS_CONFIG -> {
                val previewAndComplicationsViewHolder = viewHolder as PreviewAndComplicationsViewHolder

                previewAndComplicationsViewHolder.setDefaultComplicationDrawable()
                initializesColorsAndComplications()
            }
        }
    }

    private fun initializesColorsAndComplications() {
        val complicationIds = getComplicationIds()

        mProviderInfoRetriever.retrieveProviderInfo(
            object : OnProviderInfoReceivedCallback() {
                override fun onProviderInfoReceived(watchFaceComplicationId: Int, complicationProviderInfo: ComplicationProviderInfo?) {

                    mPreviewAndComplicationsViewHolder.updateComplicationViews(
                        if (watchFaceComplicationId == mLeftComplicationId) ComplicationLocation.LEFT else ComplicationLocation.RIGHT,
                        complicationProviderInfo,
                        storage.getComplicationColors()
                    )
                }
            },
            mWatchFaceComponentName,
            *complicationIds
        )
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == 0) {
            TYPE_PREVIEW_AND_COMPLICATIONS_CONFIG
        } else {
            TYPE_COLOR_CONFIG
        }
    }

    override fun getItemCount(): Int {
        return 2
    }

    /** Updates the selected complication id saved earlier with the new information.  */
    fun updateSelectedComplication(complicationProviderInfo: ComplicationProviderInfo?) { // Checks if view is inflated and complication id is valid.
        val selectedComplicationLocation = mSelectedComplicationLocation

        if ( selectedComplicationLocation != null ) {
            mPreviewAndComplicationsViewHolder.updateComplicationViews(
                selectedComplicationLocation,
                complicationProviderInfo,
                storage.getComplicationColors()
            )
        }
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)

        mProviderInfoRetriever.init()
    }

    override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
        super.onDetachedFromRecyclerView(recyclerView)

        // Required to release retriever for active complication data on detach.
        mProviderInfoRetriever.release()
    }

    fun updatePreviewColors() {
        mPreviewAndComplicationsViewHolder.updateComplicationsAccentColor(storage.getComplicationColors())
    }

    /**
     * Displays watch face preview along with complication locations. Allows user to tap on the
     * complication they want to change and preview updates dynamically.
     */
    class PreviewAndComplicationsViewHolder(
        view: View,
        private val mListener: ComplicationClickListener
    ) : RecyclerView.ViewHolder(view), View.OnClickListener {

        private val mLeftComplicationBackground: ImageView = view.findViewById(R.id.left_complication_background)
        private val mRightComplicationBackground: ImageView = view.findViewById(R.id.right_complication_background)
        private val mLeftComplication: ImageButton = view.findViewById(R.id.left_complication)
        private val mRightComplication: ImageButton = view.findViewById(R.id.right_complication)
        private var mDefaultComplicationDrawable: Drawable = view.context.getDrawable(R.drawable.add_complication)!!

        init {
            mLeftComplication.setOnClickListener(this)
            mRightComplication.setOnClickListener(this)
        }

        fun setDefaultComplicationDrawable() {
            mLeftComplication.setImageDrawable(mDefaultComplicationDrawable)
            mLeftComplicationBackground.visibility = View.INVISIBLE
            mRightComplication.setImageDrawable(mDefaultComplicationDrawable)
            mRightComplicationBackground.visibility = View.INVISIBLE
        }

        override fun onClick(view: View) {
            if (view == mLeftComplication) {
                mListener.onComplicationClicked(ComplicationLocation.LEFT)
            } else if (view == mRightComplication) {
                mListener.onComplicationClicked(ComplicationLocation.RIGHT)
            }
        }

        fun updateComplicationViews(location: ComplicationLocation,
                                    complicationProviderInfo: ComplicationProviderInfo?,
                                    complicationColors: ComplicationColors) {

            if (location == ComplicationLocation.LEFT) {
                updateComplicationView(
                    complicationProviderInfo, mLeftComplication,
                    mLeftComplicationBackground, complicationColors
                )
            } else if (location == ComplicationLocation.RIGHT) {
                updateComplicationView(
                    complicationProviderInfo, mRightComplication,
                    mRightComplicationBackground, complicationColors
                )
            }
        }

        private fun updateComplicationView(complicationProviderInfo: ComplicationProviderInfo?,
                                           button: ImageButton,
                                           background: ImageView,
                                           complicationColors: ComplicationColors) {

            if (complicationProviderInfo != null) {
                button.setImageIcon(complicationProviderInfo.providerIcon)
                background.visibility = View.VISIBLE
            } else {
                button.setImageDrawable(mDefaultComplicationDrawable)
                background.visibility = View.INVISIBLE
            }
            updateComplicationsAccentColor(complicationColors)
        }

        fun updateComplicationsAccentColor(colors: ComplicationColors) {
            mRightComplication.setColorFilter(colors.rightColor)
            mLeftComplication.setColorFilter(colors.leftColor)
        }
    }

    interface ComplicationClickListener {
        fun onComplicationClicked(location: ComplicationLocation)
    }

    class ColorPickerViewHolder(view: View) : RecyclerView.ViewHolder(view), View.OnClickListener {

        init {
            view.setOnClickListener(this)
        }

        override fun onClick(view: View) {
            val launchIntent = Intent(view.context, ColorSelectionActivity::class.java)
            val activity = view.context as Activity

            activity.startActivityForResult(
                launchIntent,
                ComplicationConfigActivity.UPDATE_COLORS_CONFIG_REQUEST_CODE
            )
        }
    }
}